float32 xh
float32 yh
float32 vh
float32 xr
float32 yr
float32 vr
